File: cpp/src/ch05/random_action.cc
#include <random>
#include "random_action.h"

auto mt_for_action = std::mt19937(0);

int random_action(const State &state)
{
    auto legal_actions = state.legal_actions();
    return legal_actions[mt_for_action() % legal_actions.size()];
}

void play_game(const int seed)
{
    using std::cout;
    using std::endl;
    using State = AlternateMazeState;

    auto state = State(seed);
    cout << state.to_string() << endl;

    // player
    int p = 0;
    while (!state.is_done())
    {
        cout << (p + 1) << "p -------------------" << endl;
        int action = random_action(state);
        cout << "action " << action << endl;
        state.advance(action);
        cout << state.to_string() << endl;
        p ^= 1; // same as p = (p + 1) % 2;
    }
    state.print_end_game();
}

File: rust/src/ch05/random_action.rs
use crate::base::alternate::{ActionFunc, AlternateState};
use std::sync::Arc;

use rand::{seq::SliceRandom, thread_rng};

pub fn random_action_arc<T: AlternateState>() -> ActionFunc<T> {
    Arc::new(move |state| random_action(state))
}

fn random_action<T: AlternateState>(state: &T) -> usize {
    let mut rng = thread_rng();
    let legal_actions = state.legal_actions();
    legal_actions.choose(&mut rng).unwrap().to_owned()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::base::alternate::MazeParams;
    use crate::ch05::maze_state;

    fn setup() -> maze_state::AlternateMazeState {
        let params = MazeParams {
            height: 3,
            width: 3,
            end_turn: 4,
        };
        maze_state::AlternateMazeState::new(0, params)
    }

    #[test]
    fn test_random_action() {
        let state = setup();
        let legal_actions = state.legal_actions();
        let action = random_action(&state);
        assert!(legal_actions.contains(&action));
    }
}
